File: backend/app/api/v1/expenses.py
import uuid

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.core.database import get_db
from app.models.user import User
from app.schemas.expense import ExpenseCreate, ExpenseOut, ExpenseSummaryItem, ExpenseUpdate

router = APIRouter(prefix="/expenses", tags=["expenses"])


@router.get("", response_model=list[ExpenseOut])
async def list_expenses(
    month: str | None = None,
    category_id: uuid.UUID | None = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    raise NotImplementedError


@router.post("", response_model=ExpenseOut, status_code=201)
async def create_expense(
    body: ExpenseCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    raise NotImplementedError


@router.put("/{expense_id}", response_model=ExpenseOut)
async def update_expense(
    expense_id: uuid.UUID,
    body: ExpenseUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    raise NotImplementedError


@router.delete("/{expense_id}", status_code=204)
async def delete_expense(
    expense_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    raise NotImplementedError


@router.get("/summary", response_model=list[ExpenseSummaryItem])
async def expense_summary(
    month: str | None = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    raise NotImplementedError

File: frontend/src/api/expenses.ts
import apiClient from "./client";
import type { Expense, ExpenseSummaryItem } from "../types";

export async function listExpenses(month?: string, category_id?: string): Promise<Expense[]> {
  const { data } = await apiClient.get<Expense[]>("/api/v1/expenses", { params: { month, category_id } });
  return data;
}

export async function createExpense(body: {
  category_id: string;
  amount: number;
  description: string;
  date: string;
  currency?: string;
}): Promise<Expense> {
  const { data } = await apiClient.post<Expense>("/api/v1/expenses", body);
  return data;
}

export async function updateExpense(id: string, body: Partial<Expense>): Promise<Expense> {
  const { data } = await apiClient.put<Expense>(`/api/v1/expenses/${id}`, body);
  return data;
}

export async function deleteExpense(id: string): Promise<void> {
  await apiClient.delete(`/api/v1/expenses/${id}`);
}

export async function getExpenseSummary(month?: string): Promise<ExpenseSummaryItem[]> {
  const { data } = await apiClient.get<ExpenseSummaryItem[]>("/api/v1/expenses/summary", { params: { month } });
  return data;
}
